pyfstat/logging.py

```diff
@@ -0,0 +1,136 @@
+"""
+PyFstat's logging implementation.
+
+PyFstat main logger is called `pyfstat` and can be accessed via::
+
+    import logging
+    logger = logging.getLoger('pyfstat')
+
+Basics of logging: For all our purposes, there are *logger* objects
+and *handler* objects. Loggers are the ones in charge of logging,
+hence you call them to emit a logging message with a specific logging
+level (e.g. ``logger.info``); handlers are in charge of redirecting that
+message to a specific place (e.g. a file or your terminal, which is
+usually referred to as a *stream*).
+
+The default behaviour is to attach a ``logging.StreamHandler`` to
+the *pyfstat* logger, printing out to ``sys.stdout`` upon importing the package.
+If, for any reason, ``logging`` cannot access ``sys.stdout`` at import time,
+the exception is reported via ``print`` and no handlers are attached
+(i.e. the logger won't print to ``sys.stdout``).
+
+The user can modify the logger's behaviour at run-time using ``set_up_logger``.
+This function attaches extra ``logging.StreamHandler`` and ``logging.FileHandler``
+handlers to the logger, allowing to redirect logging messages to a different
+stream or a specific output file specified using the ``outdir, label`` variables
+(with the same format as in the rest of the package).
+
+Finally, logging can be disabled, or the level changed, at run-time by
+manually configuring the *pyfstat* logger. For example, the following block of code
+will suppress logging messages below ``WARNING``:::
+
+    import logging
+    logging.getLoger('pyfstat').setLevel(logging.WARNING)
+"""
+
+import logging
+import os
+import sys
+from typing import TYPE_CHECKING, Iterable, Literal, Optional
+
+if TYPE_CHECKING:
+    import io
+
+
+def _get_default_logger() -> logging.Logger:
+    return (
+        root_logger
+        if (root_logger := logging.getLogger()).handlers
+        else set_up_logger()
+    )
+
+
+def set_up_logger(
+    outdir: Optional[str] = None,
+    label: Optional[str] = "pyfstat",
+    log_level: Literal["ERROR", "WARNING", "INFO", "DEBUG"] = "INFO",
+    streams: Optional[Iterable["io.TextIOWrapper"]] = (sys.stdout,),
+    append: bool = True,
+) -> logging.Logger:
+    """Add file and stream handlers to the `pyfstat` logger.
+
+    Handler names generated from ``streams`` and ``outdir, label``
+    must be unique and no duplicated handler will be attached by
+    this function.
+
+    Parameters
+    ----------
+    outdir:
+        Path to outdir directory. If ``None``, no file handler will be added.
+    label:
+        Label for the file output handler, i.e.
+        the log file will be called `label.log`.
+        Required, in conjunction with ``outdir``, to add a file handler.
```

Comment:
Putting this here as a random example: why the mix of single and double quotes?

Reply:
https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html

Single seems to be for italic emphasis, double is for code. Or so I understood.

Reply:
Guess we did it wrong in a bunch of other places then, but fine here. Yet another needless divergence betwen sphinx/rst and markdown...